pkg/rbd/rbd_util.go

```diff
@@ -107,64 +108,86 @@ var (
-
-	image := pOpts.RbdImageName
 	volSzMiB := fmt.Sprintf("%dM", volSz)
+	options := librbd.NewRbdImageOptions()
 
 	logMsg := "rbd: create %s size %s format 2 (features: %s) using mon %s, pool %s "
 	if pOpts.DataPool != "" {
 		logMsg += fmt.Sprintf("data pool %s", pOpts.DataPool)
+		err := options.SetString(librbd.RbdImageOptionDataPool, pOpts.DataPool)
+		if err != nil {
+			return err
+		}
 	}
 	klog.V(4).Infof(util.Log(ctx, logMsg),
-		image, volSzMiB, pOpts.ImageFeatures, pOpts.Monitors, pOpts.Pool)
+		pOpts.RbdImageName, volSzMiB, pOpts.ImageFeatures, pOpts.Monitors, pOpts.Pool)
 
-	args := []string{"create", image, "--size", volSzMiB, "--pool", pOpts.Pool, "--id", cr.ID, "-m", pOpts.Monitors, "--keyfile=" + cr.KeyFile, "--image-feature", pOpts.ImageFeatures}
-
-	if pOpts.DataPool != "" {
-		args = append(args, "--data-pool", pOpts.DataPool)
+	ioctx, err := pOpts.getIoctx(cr)
+	if err != nil {
+		return err
 	}
-	output, err := execCommand("rbd", args)
 
+	image, err := librbd.CreateWithOptions(ioctx, pOpts.RbdImageName, uint64(volSz), options)
 	if err != nil {
-		return errors.Wrapf(err, "failed to create rbd image, command output: %s", string(output))
+		return errors.Wrapf(err, "failed to create rbd image")
 	}
 
+	// image is not returned to the caller
+	// no need to Close(), the image is not open after Create()
+	_ = image
+
 	return nil
 }
 
-// rbdStatus checks if there is watcher on the image.
-// It returns true if there is a watcher on the image, otherwise returns false.
-func rbdStatus(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) (bool, string, error) {
-	var output string
-	var cmd []byte
+func (rv *rbdVolume) getConn(cr *util.Credentials) (*rados.Conn, error) {
+	args := []string{"--pool", rv.Pool, "-m", rv.Monitors, "--keyfile=" + cr.KeyFile, "--image-feature=" + rv.ImageFeatures}
+	conn, err := rados.NewConn()
+	if err != nil {
+		return nil, err
+	}
+	err = conn.ParseCmdLineArgs(args)
+	if err != nil {
+		return nil, err
+	}
 
-	image := pOpts.RbdImageName
+	err = conn.Connect()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Shutdown()
 
-	klog.V(4).Infof(util.Log(ctx, "rbd: status %s using mon %s, pool %s"), image, pOpts.Monitors, pOpts.Pool)
-	args := []string{"status", image, "--pool", pOpts.Pool, "-m", pOpts.Monitors, "--id", cr.ID, "--keyfile=" + cr.KeyFile}
-	cmd, err := execCommand("rbd", args)
-	output = string(cmd)
+	return conn, nil
+}
 
-	if err, ok := err.(*exec.Error); ok {
-		if err.Err == exec.ErrNotFound {
-			klog.Errorf(util.Log(ctx, "rbd cmd not found"))
-			// fail fast if command not found
-			return false, output, err
-		}
+func (rv *rbdVolume) getIoctx(cr *util.Credentials) (*rados.IOContext, error) {
+	conn, err := rv.getConn(cr)
+	if err != nil {
+		return nil, err
 	}
 
-	// If command never succeed, returns its last error.
+	ioctx, err := conn.OpenIOContext(rv.Pool)
 	if err != nil {
-		return false, output, err
+		return nil, err
 	}
+	defer ioctx.Destroy()
```

Comment:
This defer would be executed during the function return, thus destroying the created ioctx?

Reply:
Uhm, yes, I think so.